File: Inspect Website/InitialViewController.m
#import "InitialViewController.h"
#import "SplitViewController.h"
#import "UIHelper.h"
#import "NSAtomicNumber.h"

#include <MobileCoreServices/MobileCoreServices.h>

@interface InitialViewController ()

@property (strong, nonatomic) NSAtomicNumber * latch;
@property (strong, nonatomic) NSMutableArray * values;

@end

@implementation InitialViewController

- (void) viewDidLoad {
    [super viewDidLoad];

    [[AppState currentState] setAppearance];
    [AppState currentState].extensionContext = self.extensionContext;

    self.latch = [NSAtomicNumber numberWithInitialValue:0];
    self.values = [NSMutableArray new];

    for (NSExtensionItem *item in self.extensionContext.inputItems) {
        for (NSItemProvider *itemProvider in item.attachments) {
            NSString * urlString = [item.attributedContentText string];
            if ([itemProvider hasItemConformingToTypeIdentifier:(NSString *)kUTTypeURL]) {
                [self.latch incrementAndGet];
                // Share page from within Safari
                [itemProvider loadItemForTypeIdentifier:(NSString *)kUTTypeURL options:nil completionHandler:^(NSURL *url, NSError *error) {
                    [self.latch decrementAndGet];
                    if (url) {
                        [self.values addObject:url];
                    }
                    dispatch_async(dispatch_get_main_queue(), ^{ [self checkValues]; });
                }];
            } else if (urlString)  {
                // Share page from third-party browsers (Chrome, Brave 🦁)
                [self.values addObject:urlString];
                dispatch_async(dispatch_get_main_queue(), ^{ [self checkValues]; });
            } else if ([itemProvider hasItemConformingToTypeIdentifier:(NSString *)kUTTypePlainText]) {
                [self.latch incrementAndGet];
                // Long-press on URL in Safari or share from other app
                [itemProvider loadItemForTypeIdentifier:(NSString *)kUTTypePlainText options:nil completionHandler:^(NSString *text, NSError *error) {
                    [self.latch decrementAndGet];
                    NSURL * url = [NSURL URLWithString:text];
                    if (url) {
                        [self.values addObject:text];
                    }
                    dispatch_async(dispatch_get_main_queue(), ^{ [self checkValues]; });
                }];
            }
        }
    }
    dispatch_async(dispatch_get_main_queue(), ^{ [self checkValues]; });
}

- (void) checkValues {
    if ([self.latch getValue] <= 0) {
        if (self.values.count == 0) {
            [self closeExtension];
            return;
        }

        BOOL validURLFound = NO;

        for (id value in self.values) {
            NSURL * url;
            if ([value isKindOfClass:[NSURL class]]) {
                url = (NSURL *)value;
            } else if ([value isKindOfClass:[NSString class]]) {
                url = [NSURL URLWithString:value];
            }

            if (url) {
                if ([url.scheme isEqualToString:@"https"]) {
                    validURLFound = YES;
                    [self loadURL:url];
                }
            }
        }

        if (!validURLFound) {
            [self unsupportedURL];
        }
    }
}

- (void) parseURLString:(NSString *)urlString {
    if ([urlString hasPrefix:@"https://"]){
        [self loadURL:[NSURL URLWithString:[urlString stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]]]];
    } else {
        [self unsupportedURL];
    }
    return;
}

- (void) loadURL:(NSURL *)url {
    UIStoryboard * main = [UIStoryboard storyboardWithName:@"Main" bundle:[NSBundle mainBundle]];
    GetterTableViewController * getter = [main instantiateViewControllerWithIdentifier:@"Getter"];
    [getter presentGetter:self ForUrl:url finished:nil];
}

- (void) unsupportedURL {
    [uihelper
     presentAlertInViewController:self
     title:l(@"Unsupported Scheme")
     body:l(@"Only HTTPS sites can be inspected")
     dismissButtonTitle:l(@"Dismiss")
     dismissed:^(NSInteger buttonIndex) {
         [self closeExtension];
     }];
}

- (void) didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void) closeExtension {
    [self.extensionContext completeRequestReturningItems:self.extensionContext.inputItems completionHandler:nil];
}

- (IBAction)closeButton:(id)sender {
    [self closeExtension];
}

@end

File: Inspect Website/InitialViewController.swift
import UIKit
import MobileCoreServices
import CertificateKit

/// The initial view controller is responsible for bootstrapping the rest of the application and acting as the getter delegate.
/// This extension can be used in a number of places, and each can report the host URL in a different way.
class InitialViewController: UIViewController {
    var values: [URL] = []
    let latch = AtomicInt(defaultValue: 0)
    let requestQueue = DispatchQueue(label: "com.ecnepsnai.Inspect-Website.RequestQueue")
    var certificateChain: CKCertificateChain?
    var httpServerInfo: CKHTTPServerInfo?
    var observer: NSObjectProtocol?
    @IBOutlet weak var cancelButton: UIButton!
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!

    override func viewDidLoad() {
        super.viewDidLoad()

        if #available(iOS 13, *) {
            self.activityIndicator.style = .large
        }

        if CertificateKit.isProxyConfigured() {
            UIHelper(self).presentAlert(title: lang(key: "Proxy Detected"),
                                        body: lang(key: "proxy_warning"),
                                        dismissed: nil)
            self.closeExtension()
            return
        }

        // We use a notification to know when the user dismissed the split view
        self.observer = NotificationCenter.default.addObserver(forName: VIEW_CLOSE_NOTIFICATION,
                                                               object: nil,
                                                               queue: nil) { (_) in
            self.closeExtension()
        }

        // When you load an item from the attachment provider, it may make a HTTP request to fetch metadata
        // about that resource - we use the latch to determine if any of these potential requests may be
        // in progress when we check for URLs.
        for object in self.extensionContext?.inputItems ?? [] {
            guard let item = object as? NSExtensionItem else {
                continue
            }

            for provider in item.attachments ?? [] {
                let contentText = item.attributedContentText?.string
                // Share page from within Safari 🧭
                if provider.hasItemConformingToTypeIdentifier(kUTTypeURL as String) {
                    self.latch.increment()
                    provider.loadItem(forTypeIdentifier: (kUTTypeURL as String), options: nil) { (urlObj, _) in
                        self.latch.decrement()
                        if let url = urlObj as? URL {
                            self.values.append(url)
                            RunOnMain { self.checkValues() }
                        }
                    }
                // Share page from third-party browsers (Chrome, Brave 🦁)
                } else if let urlString = contentText {
                    if let url = URL(string: urlString) {
                        self.values.append(url)
                        RunOnMain { self.checkValues() }
                    }
                // Long-press on URL in Safari or share from other app 👆
                } else if provider.hasItemConformingToTypeIdentifier(kUTTypePlainText as String) {
                    self.latch.increment()
                    provider.loadItem(forTypeIdentifier: (kUTTypePlainText as String), options: nil) { (text, _) in
                        self.latch.decrement()
                        if let url = URL(string: text as? String ?? "") {
                            self.values.append(url)
                            RunOnMain { self.checkValues() }
                        }
                    }
                }
            }
        }
        RunOnMain { self.checkValues() }
    }
    // swiftlint:enable cyclomatic_complexity

    @IBAction func cancelButtonPressed(_ sender: UIButton) {
        self.closeExtension()
    }

    func closeExtension() {
        if let observer = self.observer {
            NotificationCenter.default.removeObserver(observer, name: VIEW_CLOSE_NOTIFICATION, object: nil)
        }
        self.extensionContext?.completeRequest(returningItems: self.extensionContext?.inputItems,
                                               completionHandler: nil)
    }

    func checkValues() {
        if self.latch.get() > 0 {
            return
        }

        if self.values.count == 0 {
            UIHelper(self).presentAlert(
            title: lang(key: "No Supported URL Found"),
            body: lang(key: "If you believe this to be in error, contact support from within the TLS Inspector app.")) {
                self.closeExtension()
            }
        }

        var foundURL = false
        for url in values {
            if url.scheme != "https" {
                continue
            }

            guard let host = url.host else {
                continue
            }

            let port = UInt16(url.port ?? 443)

            foundURL = true
            doInspect(hostAddress: host, port: port)
        }

        if !foundURL {
            UIHelper(self).presentAlert(
            title: lang(key: "No Supported URL Found"),
            body: lang(key: "Only HTTPS URLs can be inspected")) {
                self.closeExtension()
            }
        }
    }

    func doInspect(hostAddress: String, port: UInt16) {
        let parameters = UserOptions.inspectParameters(hostAddress: hostAddress)
        parameters.port = port

        let request = CKInspectRequest(parameters: parameters)
        request.execute(on: requestQueue) { oResponse, oError in
            RunOnMain {
                if let error = oError {
                    UIHelper(self).presentError(error: error) {
                        self.closeExtension()
                    }
                    return
                }
                if let response = oResponse {
                    CERTIFICATE_CHAIN = response.certificateChain
                    HTTP_SERVER_INFO = response.httpServer
                    self.performSegue(withIdentifier: "Inspect", sender: nil)
                }
            }
        }
    }
}
